File: StereoKitC/hierarchy.cpp
#include "stereokit.h"
#include "hierarchy.h"
#include "sk_math.h"
#include "libraries/array.h"

#include <DirectXMath.h> // Matrix math functions and objects
using namespace DirectX;

///////////////////////////////////////////

using namespace sk;

struct hierarchy_item_t {
	matrix   transform;
	matrix   transform_inv;
	bool32_t has_inverse;
};

struct hierarchy_state_t {
	array_t<hierarchy_item_t> stack;
	bool32_t                  enabled;
	bool32_t                  userenabled;
};
static hierarchy_state_t _local = {};
static hierarchy_state_t *local = {};

///////////////////////////////////////////

namespace sk {

///////////////////////////////////////////

void hierarchy_init() {
	local = &_local;
	*local = {};
	local->userenabled = true;
}

///////////////////////////////////////////

void hierarchy_shutdown() {
	local->stack.free();
	*local = {};
	local = nullptr;
}

///////////////////////////////////////////

void hierarchy_step() {
	if (local->stack.count > 0) {
		log_err("Render transform stack doesn't have matching begin/end calls!");
		local->stack.clear();
	}
}

///////////////////////////////////////////

void hierarchy_push(const matrix &transform, hierarchy_parent_ parent_behavior) {
	local->stack.add(hierarchy_item_t{transform, matrix_identity, false});
	local->enabled = local->userenabled;

	int32_t size = local->stack.count;
	if (size > 1 && parent_behavior == hierarchy_parent_inherit)
		matrix_mul(local->stack[size - 1].transform, local->stack[size - 2].transform, local->stack[size - 1].transform);
}

///////////////////////////////////////////

void hierarchy_pop() {
	local->stack.pop();
	if (local->stack.count == 0)
		local->enabled = false;
}

///////////////////////////////////////////

void hierarchy_set_enabled(bool32_t enabled) {
	local->userenabled = enabled;
	local->enabled     = local->stack.count != 0 && local->userenabled == (bool32_t)true;
}

///////////////////////////////////////////

bool32_t hierarchy_is_enabled() {
	return local->userenabled;
}

///////////////////////////////////////////

bool32_t hierarchy_use_top() {
	return local->enabled;
}

///////////////////////////////////////////

const matrix *hierarchy_to_world() {
	return local->enabled ? &local->stack.last().transform : &matrix_identity;
}

///////////////////////////////////////////

const matrix *hierarchy_to_local() {
	if (local->enabled) {
		hierarchy_item_t &item = local->stack.last();
		if (!item.has_inverse)
			matrix_inverse(item.transform, item.transform_inv);
		return &item.transform_inv;
	} else {
		return &matrix_identity;
	}
}

///////////////////////////////////////////

vec3 hierarchy_to_local_point(const vec3 &world_pt) {
	return matrix_transform_pt(*hierarchy_to_local(), world_pt);
}

///////////////////////////////////////////

vec3 hierarchy_to_local_direction(const vec3 &world_dir) {
	return matrix_transform_dir(*hierarchy_to_local(), world_dir);
}

///////////////////////////////////////////

quat hierarchy_to_local_rotation(const quat &world_orientation) {
	return matrix_transform_quat(*hierarchy_to_local(), world_orientation);
}

///////////////////////////////////////////

pose_t hierarchy_to_local_pose(const pose_t &world_pose) {
	return matrix_transform_pose(*hierarchy_to_local(), world_pose);
}

///////////////////////////////////////////

ray_t hierarchy_to_local_ray(ray_t world_ray) {
	return matrix_transform_ray(*hierarchy_to_local(), world_ray);
}

///////////////////////////////////////////

vec3 hierarchy_to_world_point(const vec3 &local_pt) {
	return matrix_transform_pt(*hierarchy_to_world(), local_pt);
}

///////////////////////////////////////////

vec3 hierarchy_to_world_direction(const vec3 &local_dir) {
	return matrix_transform_dir(*hierarchy_to_world(), local_dir);
}

///////////////////////////////////////////

quat hierarchy_to_world_rotation(const quat &local_orientation) {
	return matrix_transform_quat(*hierarchy_to_world(), local_orientation);
}

///////////////////////////////////////////

pose_t hierarchy_to_world_pose(const pose_t &local_pose) {
	return matrix_transform_pose(*hierarchy_to_world(), local_pose);
}

///////////////////////////////////////////

ray_t hierarchy_to_world_ray(ray_t local_ray) {
	return matrix_transform_ray(*hierarchy_to_world(), local_ray);
}

///////////////////////////////////////////

matrix hierarchy_top() {
	return local->stack.last().transform;
}

} // namespace sk

File: StereoKit/Systems/Hierarchy.cs
namespace StereoKit
{
	/// <summary>This class represents a stack of transform matrices that 
	/// build up a transform hierarchy! This can be used like an object-less 
	/// parent-child system, where you push a parent's transform onto the 
	/// stack, render child objects relative to that parent transform and 
	/// then pop it off the stack.
	/// 
	/// Performance note: if any matrices are on the hierarchy stack, any 
	/// render will cause a matrix multiplication to occur! So if you have a
	/// collection of objects with their transforms baked and cached into 
	/// matrices for performance reasons, you'll want to ensure there are no 
	/// matrices in the hierarchy stack, or that the hierarchy is disabled! 
	/// It'll save you a matrix multiplication in that case :)</summary>
	public static class Hierarchy
	{
		/// <summary>Pushes a transform Matrix onto the stack, and combines 
		/// it with the Matrix below it. Any draw operation's Matrix will now
		/// be combined with this Matrix to make it relative to the current 
		/// hierarchy. Use Hierarchy.Pop to remove it from the Hierarchy 
		/// stack! All Push calls must have an accompanying Pop call.</summary>
		/// <param name="parentTransform">The transform Matrix you want to 
		/// apply to all following draw calls.</param>
		/// <param name="parentBehavior">This determines how this matrix
		/// combines with the parent matrix below it. Normal behavior is to
		/// "inherit" the parent matrix, but there are cases where you may wish
		/// to entirely ignore the parent transform. For example, if you're in
		/// UI space, and wish to do some world space rendering.</param>
		public static void Push(in Matrix parentTransform, HierarchyParent parentBehavior = HierarchyParent.Inherit)
			=> NativeAPI.hierarchy_push(parentTransform, parentBehavior);

		/// <summary>Removes the top Matrix from the stack!</summary>
		public static void Pop()
			=> NativeAPI.hierarchy_pop();

		/// <summary>This is enabled by default. Disabling this will cause
		/// any draw call to ignore any Matrices that are on the Hierarchy
		/// stack.</summary>
		public static bool Enabled {
			get => NativeAPI.hierarchy_is_enabled();
			set => NativeAPI.hierarchy_set_enabled(value);
		}

		/// <summary>Converts a world space point into the local space of the
		/// current Hierarchy stack!</summary>
		/// <param name="worldPoint">A point in world space.</param>
		/// <returns>The provided point now in local hierarchy space!</returns>
		public static Vec3 ToLocal(Vec3 worldPoint) 
			=> NativeAPI.hierarchy_to_local_point(worldPoint);

		/// <summary>Converts a world space direction into the local space of
		/// the current Hierarchy stack! This excludes the translation
		/// component normally applied to vectors, so it's still a valid
		/// direction.</summary>
		/// <param name="worldDirection">A direction in world space.</param>
		/// <returns>The provided direction now in local hierarchy space!
		/// </returns>
		public static Vec3 ToLocalDirection(Vec3 worldDirection)
			=> NativeAPI.hierarchy_to_local_direction(worldDirection);

		/// <summary>Converts a world space rotation into the local space of
		/// the current Hierarchy stack!</summary>
		/// <param name="worldOrientation">A rotation in world space.</param>
		/// <returns>The provided rotation now in local hierarchy space!
		/// </returns>
		public static Quat ToLocal(Quat worldOrientation)
			=> NativeAPI.hierarchy_to_local_rotation(worldOrientation);


		/// <summary>Converts a world pose relative to the current
		/// hierarchy stack into local space!</summary>
		/// <param name="worldPose">A pose in world space.</param>
		/// <returns>The provided pose now in local hierarchy space!</returns>
		public static Pose ToLocal(Pose worldPose)
			=> NativeAPI.hierarchy_to_local_pose(worldPose);

		/// <summary>Converts a world ray relative to the current
		/// hierarchy stack into local space!</summary>
		/// <param name="worldRay">A ray in world space.</param>
		/// <returns>The provided ray now in local hierarchy space!</returns>
		public static Ray ToLocal(Ray worldRay)
			=> NativeAPI.hierarchy_to_local_ray(worldRay);

		/// <summary>Converts a local point relative to the current hierarchy
		/// stack into world space!</summary>
		/// <param name="localPoint">A point in local space.</param>
		/// <returns>The provided point now in world space!</returns>
		public static Vec3 ToWorld(Vec3 localPoint)
			=> NativeAPI.hierarchy_to_world_point(localPoint);

		/// <summary>Converts a local direction relative to the current 
		/// hierarchy stack into world space! This excludes the translation 
		/// component normally applied to vectors, so it's still a valid 
		/// direction.</summary>
		/// <param name="localDirection">A direction in local space.</param>
		/// <returns>The provided direction now in world space!</returns>
		public static Vec3 ToWorldDirection(Vec3 localDirection)
			=> NativeAPI.hierarchy_to_world_direction(localDirection);

		/// <summary>Converts a local rotation relative to the current 
		/// hierarchy stack into world space!</summary>
		/// <param name="localOrientation">A rotation in local space.</param>
		/// <returns>The provided rotation now in world space!</returns>
		public static Quat ToWorld(Quat localOrientation)
			=> NativeAPI.hierarchy_to_world_rotation(localOrientation);

		/// <summary>Converts a local pose relative to the current
		/// hierarchy stack into world space!</summary>
		/// <param name="localPose">A pose in local space.</param>
		/// <returns>The provided pose now in world space!</returns>
		public static Pose ToWorld(Pose localPose)
			=> NativeAPI.hierarchy_to_world_pose(localPose);

		/// <summary>Converts a local ray relative to the current
		/// hierarchy stack into world space!</summary>
		/// <param name="localRay">A ray in local space.</param>
		/// <returns>The provided ray now in world space!</returns>
		public static Ray ToWorld(Ray localRay)
			=> NativeAPI.hierarchy_to_world_ray(localRay);
	}
}
